pkg/geo/wkt/lex.go

```diff
@@ -51,13 +51,43 @@ func (e *ParseError) Error() string {
 // Constant expected by parser when lexer reaches EOF.
 const eof = 0
 
+// Stack object used for parsing the geometry type.
+// layout      is the currently parsed geometry type.
+// baseType    is a bool where true means the current context is the top-level or a base type GEOMETRYCOLLECTION.
+// mustBeEmpty is a bool used to handle the edge case where a base type geometry is allowed in a GEOMETRYCOLLECTIONM
+//             but only if it is EMPTY.
+//
+// The layout of the geometry is determined by the first geometry type keyword if it is a M, Z, or ZM variant.
+// If it is a base type geometry, the layout is determined by the number of coordinates in the first point.
+// If it is a geometrycollection, the type is the type of the first geometry in the collection.
+// Edge cases involving geometrycollections:
+// 1. GEOMETRYCOLLECTION (no type suffix) is allowed to be of type M. Normally a geometry without a type suffix
+//    is only allowed to be XY, XYZ, or XYZM.
+// 2. A base type empty geometry (e.g. POINT EMPTY) in a GEOMETRYCOLLECTIONM, GEOMETRYCOLLECTIONZ, GEOMETRYCOLLECTIONZM
+//    is permitted and takes on the type of the collection. Normally, such a geometry is XY. Note that base type
+//    non-empty geometries are still not permitted in a GEOMETRYCOLLECTIONM (e.g. POINT(0 0 0)).
+type layoutStackObj struct {
+	layout      geom.Layout
+	baseType    bool
+	mustBeEmpty bool
+}
+
 type wktLex struct {
-	line      string
-	pos       int
-	lastPos   int
-	ret       geom.T
-	curLayout geom.Layout
-	lastErr   error
+	line        string
+	pos         int
+	lastPos     int
+	ret         geom.T
+	layoutStack []layoutStackObj
+	lastErr     error
+}
+
+func makeWktLex(line string) *wktLex {
```

Comment:
nit: acronyms must be the same case, i.e. this should be makeWKTLex
nit: also, since this return a pointer, we should call this `newWKTLex` (you'll name to rename the variable below)